pkg/dependencies/dependents.go

```diff
@@ -0,0 +1,237 @@
+	// Initialize a map to hold package names mapped to their dependency nodes. This will be the returned map
+	packages := make(map[string]dependencyNode)
+
+	// Initialize maps to hold dependency and dependent edges.
+	// dependentEdges maps a package ID to the IDs of packages that depend on it (DepPkg to Pkg).
+	// dependencyEdges maps a package ID to the IDs of its dependencies (Pkg to DepPkg).
+	dependentEdges := make(map[string][]string)  // id -> dependents ids
+	dependencyEdges := make(map[string][]string) // id -> dependency ids
+
+	// Initialize a map to convert package IDs to their names.
+	idToName := make(map[string]string)
+
+	sboms, err := model.HasSBOMs(ctx, gqlClient, model.HasSBOMSpec{})
+
+	if err != nil {
+		return nil, fmt.Errorf("error getting dependencies: %v", err)
+	}
+
+	for _, resp := range sboms.HasSBOM {
+		// Skip entries from "deps.dev" because they are inconsistent.
+		if resp.Origin == deps_dev.DepsCollector {
+			continue
+		}
+		// Iterate through the included dependencies of each SBOM.
+		for _, dependency := range resp.IncludedDependencies {
+			// Construct unique names for the dependency package and the package itself.
+			// TODO: Make the names actually unique, not just add "_".
+			depPkgName := dependency.DependencyPackage.Type + "_" + dependency.DependencyPackage.Namespaces[0].Namespace + "_" + dependency.DependencyPackage.Namespaces[0].Names[0].Name
+			pkgName := dependency.Package.Type + "_" + dependency.Package.Namespaces[0].Namespace + "_" + dependency.Package.Namespaces[0].Names[0].Name
+
+			var depPkgIds []string
+			pkgId := dependency.Package.Namespaces[0].Names[0].Versions[0].Id
+
+			if len(dependency.DependencyPackage.Namespaces[0].Names[0].Versions) == 0 {
+				findMatchingDepPkgVersionIDs, err := findDepPkgVersionIDs(ctx, gqlClient, dependency.DependencyPackage.Type,
+					dependency.DependencyPackage.Namespaces[0].Namespace,
+					dependency.DependencyPackage.Namespaces[0].Names[0].Name, dependency.VersionRange)
+				if err != nil {
+					return nil, fmt.Errorf("error from FindMatchingDepPkgVersionIDs:%w", err)
+				}
+				depPkgIds = append(depPkgIds, findMatchingDepPkgVersionIDs...)
+			} else {
+				depPkgIds = append(depPkgIds, dependency.DependencyPackage.Namespaces[0].Names[0].Versions[0].Id)
+			}
+
+			for _, depPkgId := range depPkgIds {
+				// Map the IDs to their names.
+				idToName[depPkgId] = depPkgName
+				idToName[pkgId] = pkgName
+
+				// Skip "guac" files.
+				if dependency.DependencyPackage.Type == "guac" && dependency.DependencyPackage.Namespaces[0].Namespace == "files" {
+					continue
+				}
+
+				// First we need to find all the packages that have pkgName as a dependency
+
+				// Initialize a visited map and a queue for BFS to find all packages that have pkgName as a dependency.
+				visited := make(map[string]bool)
+				queue := []string{depPkgId}
+
+				// Perform BFS to mark visited nodes.
+				for len(queue) > 0 {
```

Comment:
There are two very similar graph traversals, here and on line 164. I think it would be better to factor them out to a helper function. E.g. `func reachableNodesOf(start string, edges map[string][]string) map[string]bool {...}`. Updating `packages` can be done after calling the helper function.

Reply:
I don’t think we should do this, because the readability won’t be affected that much, but we would have to preform a second loop for the following code snippet, which would further slow down the code.

``` go
for len(queue) > 0 {
	n := len(queue)

	for i := 0; i < n; i++ { // go through the entire row
		nodeId := queue[0]
		queue = queue[1:]

		node := idToName[nodeId]

		for depPkgNodeId := range visited {
			depPkgNode := idToName[depPkgNodeId]
			if _, ok := packages[depPkgNode]; !ok {
				packages[depPkgNode] = dependencyNode{dependents: make(map[string]bool)}
			}

			// Mark the node as a dependent.
			packages[depPkgNode].dependents[node] = true
		}

		// Add dependent nodes to the queue.
		queue = append(queue, dependentEdges[nodeId]...)
	}
}
```

Reply:
I think that factoring this out would make it easier to see that that some parts of this code are executed more than once on the same inputs. 

At a high level, this function finds the transitive closure of a graph of where the nodes are `Packages` and edges are dependent relationships. The input graph has the nodes and edges defined by the `IsDependency`s from a single `HasSbom`. 

One way to do this is, for every node `v` in the graph, find the reachable nodes of `v` and add to the result. On the other hand, this code does: for every edge `(v, u)`, find the reachable nodes of `u`, the nodes that reach `v`, and then add to the result. 

There are few issues with this approach. First, the graph traversals (i.e. getting reachable / reached nodes) may be done on the same node more than once, because they are done per edge, and nodes can appear in more than one edge. Secondly, for all `v`, it isn't required to find both the nodes that reach `v` and the nodes that `v` reaches, just one of these is enough to find the transitive closure. These changes would also improve readability.

Reply:
I have refactored the two graph traversals out into a function, thanks!